### PR TITLE
ARROW-9369: [Python] Support conversion from python sequence to dictionary type

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

File: cpp/src/arrow/python/python_to_arrow.cc [hunk 1]
```diff
@@ -1123,6 +1168,50 @@ class DecimalConverter : public TypedConverter<arrow::Decimal128Type, null_codin
...
+    DICTIONARY_PRIMITIVE(UINT64, UInt64Type);
+    DICTIONARY_PRIMITIVE(HALF_FLOAT, HalfFloatType);
+    DICTIONARY_PRIMITIVE(FLOAT, FloatType);
+    DICTIONARY_PRIMITIVE(DOUBLE, DoubleType);
+    DICTIONARY_PRIMITIVE(DATE32, Date32Type);
+    DICTIONARY_PRIMITIVE(DATE64, Date64Type);
+    DICTIONARY_BINARY_LIKE(BINARY, BinaryType);
+    DICTIONARY_BINARY_LIKE(STRING, StringType);
+    // DICTIONARY_BINARY_LIKE(LARGE_BINARY, LargeBinaryType);
+    // DICTIONARY_BINARY_LIKE(LARGE_STRING, LargeStringType);
```
Comment (pitrou) [inline cpp/src/arrow/python/python_to_arrow.cc:1207]: Is there a reason this isn't enabled? Is the PR unfinished?

File: python/pyarrow/scalar.pxi [hunk 2]
```diff
@@ -687,6 +687,44 @@ cdef class DictionaryScalar(Scalar):
     Concrete class for dictionary-encoded scalars.
     """
 
+    def __init__(self, index, dictionary, type):
```
Comment (jorisvandenbossche) [inline python/pyarrow/scalar.pxi:690]: Maybe we can put this in some `from_..` class method? (to keep it consistent with the other scalars that `__init__` raises)

Comment (jorisvandenbossche) [reply python/pyarrow/scalar.pxi:690]: (and can you also add some tests for this construction method?)

File: python/pyarrow/tests/test_convert_builtin.py [hunk 3]
```diff
@@ -1640,3 +1640,85 @@ def test_map_from_tuples():
...
+    pa.date32(),
+    pa.date64(),
+])
+def test_dictionary_from_integers(value_type):
+    typ = pa.dictionary(pa.int8(), value_type=value_type)
+    a = pa.array([1, 2, 1, 1, 2, 3], type=typ)
+    assert isinstance(a.type, pa.DictionaryType)
+    assert a.type.equals(typ)
+
+    expected_indices = pa.array([0, 1, 0, 0, 1, 2], type=pa.int8())
```
Comment (pitrou) [inline python/pyarrow/tests/test_convert_builtin.py:1677]: Can you test with nulls at some point?

File: python/pyarrow/tests/test_convert_builtin.py [hunk 4]
```diff
@@ -1640,3 +1640,85 @@ def test_map_from_tuples():
...
+
+    a = pa.array(range(2**7 + 1), type=typ)
+    expected = pa.dictionary(pa.int16(), pa.int64())
+    assert a.type.equals(expected)
+
+
+def test_dictionary_from_strings():
+    for value_type in [pa.binary(), pa.string()]:
+        typ = pa.dictionary(pa.int8(), value_type)
+        a = pa.array(["", "a", "bb", "a", "bb", "ccc"], type=typ)
```
Comment (pitrou) [inline python/pyarrow/tests/test_convert_builtin.py:1707]: Also test with nulls?

File: python/pyarrow/tests/test_scalars.py [hunk 5]
```diff
@@ -550,8 +550,8 @@ def test_dictionary():
         with pytest.warns(FutureWarning):
             assert s.dictionary_value.as_py() == v
 
-    with pytest.raises(pa.ArrowNotImplementedError):
-        pickle.loads(pickle.dumps(s))
+    restored = pickle.loads(pickle.dumps(s))
+    assert restored.equals(s)
```
Comment (pitrou) [inline python/pyarrow/tests/test_scalars.py:554]: This should be inside the loop.
(also, null values are not tested?)

**Comment by kszucs** [comment]
We can close this in favor of #8088

**Pull request closed without merging**